src/Interpreters/InterpreterCreateIndexQuery.cpp

```diff
@@ -38,7 +38,7 @@ BlockIO InterpreterCreateIndexQuery::execute()
     query_ptr->as<ASTCreateIndexQuery &>().setDatabase(table_id.database_name);
 
     DatabasePtr database = DatabaseCatalog::instance().getDatabase(table_id.database_name);
-    if (typeid_cast<DatabaseReplicated *>(database.get()) && !current_context->getClientInfo().is_replicated_database_internal)
+    if (database->shouldReplicateQuery(getContext(), query_ptr))
```

Comment:
I get that it's cleaner but to me, the assumption that "if this passes it's a DatabaseReplicated" isn't that obvious.
What about having 
```
if (auto * replicated = typeid_cast<DatabaseReplicated *>(database.get());
    replicated && replicated->shouldReplicateQuery())
    ....
```
and not have `shouldReplicateQuery` as virtual?

Reply:
Maybe just rename `shouldReplicateQuery` to something more clear like `shouldExecuteThroughDistributedDDLQueue` and make `tryEnqueueReplicatedDDL` virtual too?

Reply:
> make `tryEnqueueReplicatedDDL` virtual too

This is also okay with me, no need for rename with it.
I just didn't like the mix of having check virtual and then the actual enqueue requires cast.